src/storage/src/scheduler.rs

```diff
@@ -537,10 +538,49 @@ mod tests {
             .unwrap();
     }
 
+    struct MockAsyncHandler<F> {
+        cb: F,
+    }
+
+    #[async_trait::async_trait]
+    impl<F> Handler for MockAsyncHandler<F>
+    where
+        F: Fn() -> BoxFuture<'static, ()> + Send + Sync,
+    {
+        type Request = MockRequest;
+
+        async fn handle_request(
+            &self,
+            _req: Self::Request,
+            token: BoxedRateLimitToken,
+            finish_notifier: Arc<Notify>,
+        ) -> Result<()> {
+            let fut = (self.cb)();
+            fut.await;
+            token.try_release();
+            finish_notifier.notify_one();
+            Ok(())
+        }
+    }
+
     #[tokio::test]
     async fn test_schedule_duplicate_tasks() {
         common_telemetry::init_default_ut_logging();
-        let handler = MockHandler { cb: || {} };
+        let (tx, rx) = tokio::sync::watch::channel(false);
+        let handler = MockAsyncHandler {
+            cb: move || {
+                let mut rx = rx.clone();
+                Box::pin(async move {
+                    // Block the handler so it can't handle more requests.
+                    loop {
+                        rx.changed().await.unwrap();
+                        if *rx.borrow() {
+                            break;
+                        }
+                    }
```

Comment:
Looks like you invent a [`Barrier`](https://docs.rs/tokio/latest/tokio/sync/struct.Barrier.html)

Reply:
The handler will trigger the callback twice instead of 10 times. I need a mark to block the callback instead of a barrier to block all tasks.